### PR TITLE
WIP: Introduce txt-prov for provisioning a tpm2 from userspace

### PR DESCRIPTION
**Pull request by ChriMarMe**
Signed-off-by: Christopher Meis <christopher.meis@9elements.com>

### CODE REVIEW

File: cmd/txt-prov/cmd.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	fmt.Println("BSD 3-Clause License")
+	fmt.Println("")
+	fmt.Println("Copyright (c) 2020, 9elements Agency GmbH.")
+	fmt.Println("Copyright (c) 2020, facebook Inc.")
+	fmt.Println("All rights reserved.")
+}
+
+func showHelp() {
+	fmt.Println("Converged Security Suite TXT provisioning tool")
+	fmt.Println("Usage: prov-tools [-lf][-d -pwWrite][-pwWrite][-pwDelete][-h shaXXX][-w][-d][-ps][-h][-v]")
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:38]: only optional arguments? Make at least on (the last or first) mandatory.

Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:38]: -ps isn't explained below

File: cmd/txt-prov/cmd.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	fmt.Println("\t-dAUX [-pwDelete / -lf] : Overwrites PS index with AUXdelete bit set. Needs a reboot afterwards.")
+	fmt.Println("\t-dPS [-pwDelete / -lf : Delete PS index. Only permitted after using -dAUX and a reboot.")
+	fmt.Println("\t-h SHA256,SHA384,SHA512 : Select list of hash algorithms for creating the policy. Supported: SHA1(TPM 1.2 only), SHA256(TPM 2.0 default), SHA384, SHA512")
+	fmt.Println("\t-lf : Loads files named 'del.<HashAlg>', 'write.<HashAlg>', 'psPolicy.<HashAlg>' from the working directory")
+	fmt.Println("\t-pWrite password : Give a passwort for write branch of PS policy")
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:43]: is password an argument? If so the usage  above must be [-pWrite password]

File: cmd/txt-prov/cmd.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	fmt.Println("\t-pDelete password : Give a passwort for delete branch of PS policy")
+	fmt.Println("\t-s : Select if policies shall be safen. Default: false")
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:45]: safen?

File: cmd/txt-prov/main.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	"fmt"
+
+	tss "github.com/9elements/go-tss"
+)
+
+func main() {
+	flag.Parse()
+	err := run()
+	if err != nil {
+		fmt.Printf("%v", err)
```
Comment (PatrickRudolph) [inline cmd/txt-prov/main.go:15]: set os.Exit(1) on error or use panic()

File: cmd/txt-prov/cmd.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	var writeHash []byte
+	var psPolicy []byte
+	files, err := ioutil.ReadDir(".")
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+	for _, file := range files {
+		var data []byte
+		if strings.HasSuffix(file.Name(), ".SHA256") {
+			*hAlg = crypto.SHA256
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:95]: hAlg = &crypto.SHA256?

File: cmd/txt-prov/cmd.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+			data = make([]byte, crypto.SHA256)
+		}
```
Comment (PatrickRudolph) [inline cmd/txt-prov/cmd.go:97]: else {
continue
}

File: cmd/txt-prov/psdel.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
+package main
```
Comment (PatrickRudolph) [inline cmd/txt-prov/psdel.go:1]: move to pkg/provisioning ?

File: cmd/txt-prov/main.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+			}
+			if err := writeFile("delHash", delHash, *hA); err != nil {
+				return err
+			}
+			if err = writeFile("writeHash", writeHash, *hA); err != nil {
+				return err
+			}
+		}
+
+		if !*deletePS && !*deleteAUX {
```
Comment (PatrickRudolph) [inline cmd/txt-prov/main.go:68]: looks like the 3 possible commands are DELETEPS, DELETEAUX, PROVISION. That should be the last argument you pass to this application and it must be mandatory.

File: cmd/txt-prov/cmd.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChriMarMe) [reply cmd/txt-prov/cmd.go:38]: This only prints the small help info from the flag.X() description but is not detailed enough i think.

File: cmd/txt-prov/cmd.go (hunk 5 above)
Comment (ChriMarMe) [reply cmd/txt-prov/cmd.go:95]: Nope, not in Golang

File: cmd/txt-prov/cmd.go (hunk 3 above)
Comment (ChriMarMe) [reply cmd/txt-prov/cmd.go:45]: Upsi. Typo :D

File: cmd/txt-prov/cmd.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChriMarMe) [reply cmd/txt-prov/cmd.go:38]: This does not give enough information.

**Pull request merged**